File: frontend/lib/trace/engine/distance.ts
import { GrayF, Mask } from './buffers';

/**
 * Exact Euclidean distance transform. See ALGORITHMS.md §9.
 *
 * Felzenszwalb & Huttenlocher's two-pass 1-D lower-envelope algorithm: O(n) per row and per column,
 * and **exact**, not the 3-4 chamfer approximation. Stroke width comes straight off this
 * (`width = 2 * distance` at the skeleton), and a chamfer's 6% directional error reads as a visible
 * wobble in a width-modulated stroke.
 */

// Large enough to stand in for infinity in the squared domain without overflowing a double's exact
// integer range, so the parabola intersections stay well-conditioned.
const INF = 1e20;

/**
 * @param src              the binary image
 * @param insideForeground true (default) measures each pixel's distance to the nearest **background**
 *                         pixel, so the values are non-zero inside blobs — this is what stroke width
 *                         needs. false measures the distance to the nearest foreground pixel.
 * @returns a new GrayF of Euclidean distances in pixels. When there is no source pixel at all (an
 *          all-foreground mask with `insideForeground`), every value is capped at the image diagonal
 *          rather than left at infinity, so downstream arithmetic stays finite.
 */
export function euclidean(src: Mask, insideForeground = true): GrayF {
  const w = src.width;
  const h = src.height;
  const n = w * h;
  const d = src.data;
  const f = new Float64Array(n);
  let anySource = false;
  for (let i = 0; i < n; i++) {
    const isSource = insideForeground ? d[i] === 0 : d[i] !== 0;
    if (isSource) {
      f[i] = 0;
      anySource = true;
    } else {
      f[i] = INF;
    }
  }
  if (!anySource) {
    // Every distance would be infinite. Zeros are the honest answer and they keep downstream arithmetic
    // finite; an image-diagonal placeholder reads as real structure to the stroke-width sampler and
    // would put a uniform fat stroke on a mask that contains nothing.
    return new GrayF(w, h);
  }

  const maxDim = Math.max(w, h);
  const row = new Float64Array(maxDim);
  const dt = new Float64Array(maxDim);
  const v = new Int32Array(maxDim);
  const z = new Float64Array(maxDim + 1);

  // Columns first, then rows: the order is irrelevant to the answer but doing columns first keeps the
  // row pass reading contiguous memory, which is measurably faster at working resolution.
  for (let x = 0; x < w; x++) {
    for (let y = 0; y < h; y++) row[y] = f[y * w + x];
    lowerEnvelope(row, h, dt, v, z);
    for (let y = 0; y < h; y++) f[y * w + x] = dt[y];
  }
  for (let y = 0; y < h; y++) {
    const base = y * w;
    for (let x = 0; x < w; x++) row[x] = f[base + x];
    lowerEnvelope(row, w, dt, v, z);
    for (let x = 0; x < w; x++) f[base + x] = dt[x];
  }

  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) out[i] = Math.sqrt(f[i]);
  return new GrayF(w, h, out);
}

/** 1-D squared-distance transform of a sampled function, writing `n` values into `dt`. */
function lowerEnvelope(
  fRow: Float64Array,
  n: number,
  dt: Float64Array,
  v: Int32Array,
  z: Float64Array,
): void {
  let k = 0;
  v[0] = 0;
  z[0] = -INF;
  z[1] = INF;
  for (let q = 1; q < n; q++) {
    let top = k;
    let s = 0;
    for (;;) {
      const p = v[top];
      s = (fRow[q] + q * q - (fRow[p] + p * p)) / (2 * (q - p));
      // `top > 0` rather than testing z[0] = -INF: with 1e20 standing in for infinity the comparison
      // against z[0] is arithmetically true but not *guaranteed* true, and popping past the base of the
      // envelope reads v[-1].
      if (top > 0 && s <= z[top]) {
        top--;
        continue;
      }
      break;
    }
    k = top + 1;
    v[k] = q;
    z[k] = s;
    z[k + 1] = INF;
  }
  let k2 = 0;
  for (let q = 0; q < n; q++) {
    while (z[k2 + 1] < q) k2++;
    const p = v[k2];
    const diff = q - p;
    dt[q] = diff * diff + fRow[p];
  }
}

/**
 * Stroke width from a distance transform: twice the distance to the nearest background pixel.
 * @returns `2 * dt(x, y)`, with the coordinates edge-clamped so a caller sampling at a path vertex that
 *          rounds one pixel outside the image gets the border value instead of NaN.
 */
export function strokeWidthAt(dt: GrayF, x: number, y: number): number {
  return 2 * dt.clamped(x | 0, y | 0);
}

File: android/core-imaging/src/main/java/com/offlinetracer/imaging/Distance.kt
package com.offlinetracer.imaging

import kotlin.math.max
import kotlin.math.sqrt

/**
 * Exact Euclidean distance transform.
 *
 * Felzenszwalb & Huttenlocher's two-pass 1-D lower-envelope algorithm: O(n) per row and per column,
 * and **exact**. The usual 3-4 or 5-7-11 chamfer approximation is not good enough here because the
 * only consumer is stroke-width estimation (`2 × distance at the skeleton`) — a chamfer's few
 * percent of anisotropic error becomes a stroke that visibly changes width as it turns a corner,
 * which is precisely the artefact width modulation exists to avoid.
 */
object Distance {

    /**
     * Stand-in for infinity. A real `POSITIVE_INFINITY` propagates `INF - INF = NaN` through the
     * parabola intersection whenever two unreachable columns are compared, and the NaN then poisons
     * the envelope for the rest of the row. This value is finite, so the intersection degenerates
     * to the harmless midpoint, and it is far larger than any squared distance an image can produce
     * (a 2^15-pixel edge gives ~2.1e9).
     */
    private const val FAR = 1.0e20

    /**
     * Euclidean distance (not squared) from every pixel to the nearest pixel of the opposite class,
     * as a [GrayF] in pixel units.
     *
     * With [insideForeground] true — the default and what stroke width needs — foreground pixels
     * hold their distance to the nearest background pixel and background pixels are 0. With it
     * false the roles swap, giving the distance from empty space to the nearest ink.
     *
     * Out-of-bounds is *not* treated as background here, unlike the mask/morphology convention: the
     * transform is defined on the array it is given, so a shape that runs off the edge keeps
     * growing rather than being cut by an imaginary border. A mask with no pixels of the reference
     * class at all returns all zeros, since every distance would otherwise be infinite.
     */
    fun euclidean(src: Mask, insideForeground: Boolean = true): GrayF {
        val w = src.width
        val h = src.height
        val n = w * h
        val d = src.data
        val f = DoubleArray(n)
        var hasSource = false
        for (i in 0 until n) {
            val isSource = if (insideForeground) !d[i] else d[i]
            if (isSource) {
                hasSource = true
            } else {
                f[i] = FAR
            }
        }
        if (!hasSource) return GrayF(w, h)

        val g = DoubleArray(n)
        val m = max(w, h)

        // Column pass over columns, row pass over rows, with a barrier between them — the row pass
        // reads every element of `g`, so it cannot start until the last column has been written, and
        // `Parallel.chunks` returning is that barrier.
        //
        // **Each share allocates its own `v`/`z`.** Hoisting them out here, the way the single-threaded
        // version did, is the classic way to break a parallel Felzenszwalb transform: `v` is the list of
        // vertices of the lower envelope and `z` the breakpoints between them, and the second half of
        // [dt1d] reads back the list the first half built. A neighbouring share overwriting entry `k`
        // in between substitutes another line's parabola, and the result is not garbage — it is a
        // plausible-looking distance field that is quietly wrong, which is the worst kind. The
        // allocation is O(threads) per call, not O(pixels).
        Parallel.chunks(w, Parallel.ROWS_KERNEL) { fromX, toX ->
            val v = IntArray(m)
            val z = DoubleArray(m + 1)
            for (x in fromX until toX) dt1d(f, x, w, g, x, w, h, v, z)
        }
        Parallel.rows(h, Parallel.ROWS_KERNEL) { fromY, toY ->
            val v = IntArray(m)
            val z = DoubleArray(m + 1)
            for (y in fromY until toY) dt1d(g, y * w, 1, f, y * w, 1, w, v, z)
        }

        val out = FloatArray(n)
        Parallel.chunks(n, Parallel.PIXELS_MAP) { from, to ->
            for (i in from until to) out[i] = sqrt(f[i]).toFloat()
        }
        return GrayF(w, h, out)
    }

    /**
     * Stroke width implied by a distance transform at `(x, y)`: `2 × dt`, since the distance at the
     * centreline is the half-width. Coordinates are edge-clamped, so a caller walking a polyline
     * that ends exactly on the border still gets a usable width instead of an exception.
     */
    fun strokeWidthAt(dt: GrayF, x: Int, y: Int): Float = 2f * dt.clamped(x, y)

    /**
     * Lower envelope of the parabolas `(q - i)² + f[i]` along one strided line.
     *
     * [f] and [out] must be different arrays: the second loop reads `f` at the vertex of the
     * winning parabola, which can lie *ahead* of the position being written, so writing in place
     * corrupts values that have not been read yet.
     *
     * [v] and [z] are caller-supplied scratch and must belong to **one thread**. They carry state from
     * the first loop to the second, so two lines sharing them is a correctness bug, not a performance
     * one. See the allocation site in [euclidean].
     */
    private fun dt1d(
        f: DoubleArray, fOff: Int, fStride: Int,
        out: DoubleArray, oOff: Int, oStride: Int,
        n: Int, v: IntArray, z: DoubleArray,
    ) {
        if (n <= 0) return
        var k = 0
        v[0] = 0
        z[0] = -FAR
        z[1] = FAR
        for (q in 1 until n) {
            val fq = f[fOff + q * fStride] + q.toDouble() * q
            var vk = v[k]
            var s = (fq - (f[fOff + vk * fStride] + vk.toDouble() * vk)) / (2.0 * q - 2.0 * vk)
            while (k > 0 && s <= z[k]) {
                k--
                vk = v[k]
                s = (fq - (f[fOff + vk * fStride] + vk.toDouble() * vk)) / (2.0 * q - 2.0 * vk)
            }
            k++
            v[k] = q
            z[k] = s
            z[k + 1] = FAR
        }
        k = 0
        for (q in 0 until n) {
            while (z[k + 1] < q) k++
            val vk = v[k]
            val dq = (q - vk).toDouble()
            out[oOff + q * oStride] = dq * dq + f[fOff + vk * fStride]
        }
    }
}
